### PR TITLE
[ROX-12212] : Optimize node and related GraphQL resolvers; Avoid pulling full node with scan and vulns by default in node datastore

### PR DESCRIPTION
**Pull request by charmik-redhat**
## Description

All below improvements are for Postgres.
- In node datastore and postgres store, add ability to avoid pulling full node with scan, components and vulns.
- In Node GraphQL, pull partial node excluding scan, until `Scan()` sub resolver is called, when full node is retrieved with scan, components and vulns.
- Cache embedded node scan component and other useful info in context and use that whenever possible instead of querying datastores.
- Cache embedded node scan vulns in context and use that when possible instead of querying datastores.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

- Added unit tests and benchmark tests for node datastore and NodeScan GraphQL.
- Performed manual testing by calling nodes->nodeScan->nodeComponents->nodeVulnerabilities resolvers to verify the correctness of results.

### node_scan_benchmark_test results
```
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkNodeResolver
BenchmarkNodeResolver/GetNodeComponentsInNodeScanResolver
BenchmarkNodeResolver/GetNodeComponentsInNodeScanResolver-16         	      30	  40213754 ns/op
BenchmarkNodeResolver/GetNodeComponentsWithoutNodeScanResolver
BenchmarkNodeResolver/GetNodeComponentsWithoutNodeScanResolver-16    	       6	 179263362 ns/op
BenchmarkNodeResolver/GetNodeComponentsDerivedFieldsWithNodeScanResolver
BenchmarkNodeResolver/GetNodeComponentsDerivedFieldsWithNodeScanResolver-16         	      15	  85106791 ns/op
BenchmarkNodeResolver/GetNodeComponentsDerivedWithoutNodeScanResolver
BenchmarkNodeResolver/GetNodeComponentsDerivedWithoutNodeScanResolver-16            	       1	1384612231 ns/op
BenchmarkNodeResolver/GetNodeOnly
BenchmarkNodeResolver/GetNodeOnly-16                                                	    3465	    325572 ns/op
BenchmarkNodeResolver/GetNodeWithCounts
BenchmarkNodeResolver/GetNodeWithCounts-16                                          	      39	  28955174 ns/op
PASS
```

### Node datastore benchmark test results comparing GetNodesBatch (includes full node scan) and GetManyNodeMetadata (does not include node scan)
```
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkGetManyNodes
BenchmarkGetManyNodes/GetNodesBatch
BenchmarkGetManyNodes/GetNodesBatch-16         	      15	  98662333 ns/op
BenchmarkGetManyNodes/GetManyNodeMetadata
BenchmarkGetManyNodes/GetManyNodeMetadata-16   	    2220	    696879 ns/op
PASS
```

### CODE REVIEW

File: central/node/datastore/store/postgres/store.go [hunk 1]
```diff
@@ -917,10 +917,68 @@ func (s *storeImpl) GetNodeMetadata(ctx context.Context, id string) (*storage.No
...
+		}
+	}
+	return elems, missingIndices, nil
+}
+
 //// Used for testing
 
 // CreateTableAndNewStore returns a new Store instance for testing
-func CreateTableAndNewStore(ctx context.Context, t *testing.T, db *pgxpool.Pool, gormDB *gorm.DB, noUpdateTimestamps bool) Store {
+func CreateTableAndNewStore(ctx context.Context, db *pgxpool.Pool, gormDB *gorm.DB, noUpdateTimestamps bool) Store {
```
Comment (keyallis) [inline central/node/datastore/store/postgres/store.go:981]: What's the reason for this change? I don't see a similar change in the Image equivalent PR

Comment (charmik-redhat) [reply central/node/datastore/store/postgres/store.go:981]: I removed that because testing.T argument was not used in the function, but now I think that the reason it was there was to ensure that the function gets called only during testing. I would revert the change.

Comment (dashrews78) [reply central/node/datastore/store/postgres/store.go:981]: looks like this was the only store that still had the testing reference.  So I presume that was a remnant from when the store was originally generated and before it diverged from the auto-generated code.  So the testing reference should probably be removed.

File: central/graphql/resolvers/embeddedobjs/node_component_context.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	value := context.Value(nodeComponentContextKey{})
+	if value == nil {
+		return nil
+	}
+	return value.(*nodeComponentContextValue).lastScanned
+}
+
+// NodeComponentOSFromContext returns the operating system of the component, scoped to embedding node, from the input context.
+func NodeComponentOSFromContext(context context.Context) string {
```
Comment (dashrews78) [inline central/graphql/resolvers/embeddedobjs/node_component_context.go:55]: This doesn't seem to be used, so do we even need to store the os in the context?

File: central/graphql/resolvers/embeddedobjs/node_vuln_context.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/stackrox/rox/generated/storage"
+)
+
+// nodeVulnContextKey is the key for the *storage.NodeVulnerability value in the context.
+type nodeVulnContextKey struct{}
+
+// nodeVulnContextValue holds the value of the distro in the context.
+type nodeVulnContextValue struct {
+	vuln        *storage.NodeVulnerability
+	lastScanned *types.Timestamp
```
Comment (dashrews78) [inline central/graphql/resolvers/embeddedobjs/node_vuln_context.go:16]: It does not look like you are ever setting or getting this field.  Do we need it?

File: central/graphql/resolvers/node_components.go [hunk 4]
```diff
@@ -313,10 +340,71 @@ func (resolver *nodeComponentResolver) TopNodeVulnerability(ctx context.Context)
...
+			if err != nil {
+				return nil, err
+			}
+			resolver.ctx = embeddedobjs.NodeVulnContext(ctx, vuln)
+			idToVals[id] = resolver
+		}
+	}
+
+	// For now, sort by ID.
+	resolvers := make([]*nodeCVEResolver, 0, len(idToVals))
```
Comment (dashrews78) [inline central/graphql/resolvers/node_components.go:393]: nit:  I think naming these so there is more distinction between resolvers and resolverI would be useful.  Also resolverI should probably be plural.

File: central/graphql/resolvers/node_scan.go [hunk 5]
```diff
@@ -29,5 +37,55 @@ func (resolver *nodeScanResolver) NodeComponentCount(_ context.Context, args Raw
...
+			if err != nil {
+				return nil, err
+			}
+			resolver.ctx = embeddedobjs.NodeComponentContext(ctx, os, nodeScan.GetScanTime(), embeddedComponent)
+			idToComponent[id] = resolver
+		}
+	}
+
+	// For now, sort by IDs.
+	resolvers := make([]*nodeComponentResolver, 0, len(idToComponent))
```
Comment (dashrews78) [inline central/graphql/resolvers/node_scan.go:74]: nit: I think naming these so there is more distinction between resolvers and resolverI would be useful. Also resolverI should probably be plural.

File: central/graphql/resolvers/nodes.go [hunk 6]
```diff
@@ -523,7 +527,23 @@ func (resolver *nodeResolver) PlottedNodeVulnerabilities(ctx context.Context, ar
...
+	scan := resolver.data.GetScan()
+	if features.PostgresDatastore.Enabled() {
+		// If scan is pulled, it is most likely to fetch all components and vulns contained in node.
+		// Therefore, load the node again with full scan.
+		nodeLoader, err := loaders.GetNodeLoader(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		// If Postgres is not enabled, node loader always pulls full node.
```
Comment (dashrews78) [inline central/graphql/resolvers/nodes.go:539]: Not sure I'm following this comment since it is in the Postgres block

File: central/graphql/resolvers/testutils.go [hunk 7]
```diff
@@ -46,7 +63,7 @@ func setupPostgresConn(t testing.TB) (*pgxpool.Pool, *gorm.DB) {
 	return pool, gormDB
 }
 
-func setupResolver(
+func setupResolverForImageGraphQLTestsWithPostgres(
```
[the remark this answers is not in the document]
Comment (dashrews78) [reply central/graphql/resolvers/testutils.go:66]: agreed that name is really long.

File: central/node/datastore/store/dackbox/store_impl.go [hunk 8]
```diff
@@ -112,7 +114,12 @@ func (b *storeImpl) GetNodeMetadata(_ context.Context, id string) (*storage.Node
 	return node, node != nil, err
 }
 
-// GetNodesBatch returns nodes with given ids.
+func (b *storeImpl) GetManyNodeMetadata(ctx context.Context, id []string) ([]*storage.Node, []int, error) {
```
Comment (dashrews78) [inline central/node/datastore/store/dackbox/store_impl.go:117]: you should add the function comment to be consistent.

File: central/node/datastore/store/postgres/store.go [hunk 9]
```diff
@@ -67,8 +66,9 @@ type Store interface {
 	Delete(ctx context.Context, id string) error
 	GetIDs(ctx context.Context) ([]string, error)
 	GetMany(ctx context.Context, ids []string) ([]*storage.Node, []int, error)
-	// GetNodeMetadata gets the node without scan/component data.
+	// GetNodeMetadata and GetManyNodeMetadata returns the node without scan/component data.
```
Comment (dashrews78) [inline central/node/datastore/store/postgres/store.go:69]: nit:  my preference would be to put the comment on each function instead of grouping them.  But I understand why you are doing it this way.  So I leave it to you.

**Review by dashrews78 — CHANGES_REQUESTED** [review]
I didn't see anything super major, but some stuff you should clean up.  Additionally, I think I'd like to see it after a rebase since this PR is pretty stale.

File: central/graphql/resolvers/testutils.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (charmik-redhat) [reply central/graphql/resolvers/testutils.go]: Yeah, changed the name

File: central/graphql/resolvers/embeddedobjs/node_component_context.go (hunk 2 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/embeddedobjs/node_component_context.go]: Ah good catch, I removed the function

File: central/graphql/resolvers/embeddedobjs/node_vuln_context.go (hunk 3 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/embeddedobjs/node_vuln_context.go]: Yeah, removed it

File: central/graphql/resolvers/nodes.go (hunk 6 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/nodes.go]: Ah yes, the comment should be before the Postgres block.

File: central/graphql/resolvers/loaders/nodes.go [hunk 10]
```diff
@@ -58,16 +60,38 @@ type nodeLoaderImpl struct {
...
+	if err != nil {
+		return nil, err
+	}
+	// Load the full node if full scan is not available.
+	if node.GetComponents() == 0 || len(node.GetScan().GetComponents()) > 0 {
+		return node, nil
+	}
+
+	ndl.lock.Lock()
+	delete(ndl.loaded, id)
```
Comment (c-du) [inline central/graphql/resolvers/loaders/nodes.go:91]: I did not see codes to expire entries in loaded, will it become full node eventually? Of course, that is not introduced in this PR.

File: central/graphql/resolvers/node_scan_benchmark_test.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
+package resolvers
```
Comment (c-du) [inline central/graphql/resolvers/node_scan_benchmark_test.go:1]: Can you post some benchmark results here?

File: central/graphql/resolvers/loaders/nodes.go [hunk 12]
```diff
@@ -58,16 +60,38 @@ type nodeLoaderImpl struct {
...
 		return nil, err
 	}
 	return nodes, nil
 }
 
-// FromID loads a node from an ID.
+// FromID loads a node from an ID. Does not load scan components and vulnerabilities when Postgres is enabled.
 func (ndl *nodeLoaderImpl) FromID(ctx context.Context, id string) (*storage.Node, error) {
-	nodes, err := ndl.load(ctx, []string{id})
+	nodes, err := ndl.load(ctx, []string{id}, false)
```
Comment (c-du) [inline central/graphql/resolvers/loaders/nodes.go:72]: It would be helpful if we have metrics around load cache here. But this is out of scope for this PR.

**Review by c-du — APPROVED** [review]
LGTM overall, just a few questions.

File: central/graphql/resolvers/embeddedobjs/node_component_context.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"context"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/stackrox/rox/generated/storage"
+)
+
+// nodeComponentContextKey is the key for the *storage.EmbeddedNodeScanComponent value in the context.
+// All data is scoped to the embedding node.
+type nodeComponentContextKey struct{}
```
[the remark this answers is not in the document]
Comment (charmik-redhat) [reply central/graphql/resolvers/embeddedobjs/node_component_context.go:12]: We use uniquely named structs as keys within contexts. Also, context doesn't recommend using string or any builtin type as keys.

File: central/graphql/resolvers/loaders/nodes.go (hunk 10 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/loaders/nodes.go:91]: Yes it will become full node eventually, and the next time same node it queried, it will not fetch full node again.

File: central/graphql/resolvers/node_scan_benchmark_test.go (hunk 11 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/node_scan_benchmark_test.go:1]: Do I add it in comment here? Or in the top of file?

File: central/graphql/resolvers/loaders/nodes.go (hunk 12 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/loaders/nodes.go:72]: Yeah, it would be useful to know especially memory usages, to see if the cache takes up too much space.

**Review by dashrews78 — APPROVED** [review]
LGTM

File: central/graphql/resolvers/node_components.go [hunk 14]
```diff
@@ -246,7 +304,22 @@ func (resolver *nodeComponentResolver) NodeCount(ctx context.Context, args RawQu
...
-	return resolver.root.NodeVulnerabilities(resolver.nodeComponentScopeContext(ctx), args)
+
+	if resolver.ctx == nil {
+		resolver.ctx = ctx
+	}
+
+	// Short path. Full node is embedded when node scan resolver is called.
+	embeddedComponent := embeddedobjs.NodeComponentFromContext(resolver.ctx)
+	if embeddedComponent == nil || !env.PostgresDatastoreEnabled.BooleanSetting() {
+		return resolver.root.NodeVulnerabilities(resolver.nodeComponentScopeContext(ctx), args)
```
Comment (keyallis) [inline central/graphql/resolvers/node_components.go:315]: Should this line be using `ctx` or `resolver.ctx`?

File: central/graphql/resolvers/node_scan.go [hunk 15]
```diff
@@ -29,5 +42,52 @@ func (resolver *nodeScanResolver) NodeComponentCount(_ context.Context, args Raw
 }
 
 func (resolver *nodeScanResolver) NodeComponents(_ context.Context, args PaginatedQuery) ([]NodeComponentResolver, error) {
```
Comment (keyallis) [inline central/graphql/resolvers/node_scan.go:44]: Do we need the:
```go
if resolver.ctx == nil {
resolver.ctx = ctx
}
```

File: central/graphql/resolvers/node_scan_benchmark_test.go (hunk 11 above)
Comment (keyallis) [reply central/graphql/resolvers/node_scan_benchmark_test.go:1]: If you could include it in the PR description

File: central/graphql/resolvers/nodes.go [hunk 16]
```diff
@@ -546,3 +570,15 @@ func (resolver *nodeResolver) nodeScopeContext(ctx context.Context) context.Cont
 		ID:    resolver.data.GetId(),
 	})
 }
+
+//// Node scan-related fields pulled as direct sub-resolvers of node.
+
+func (resolver *nodeResolver) ScanNotes(ctx context.Context) []string {
+	value := resolver.data.GetScan().GetNotes()
+	return stringSlice(value)
```
Comment (keyallis) [inline central/graphql/resolvers/nodes.go:578]: nit: this can be a one line, same with `ScanTime`

**Review by keyallis — COMMENTED** [review]
Couple of comments

File: central/graphql/resolvers/node_components.go (hunk 14 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/node_components.go:315]: No it should still use ctx because `resolver.nodeComponentScopeContext` func also does below assignment
```
if resolver.ctx == nil {
 		resolver.ctx = ctx
 	}
```
It is redundant in this case, but we need non-nil resolver.ctx when calling `getNodeCVEResolvers` on ln 322 below. resolver.nodeComponentScopeContext get called only within the if condition, so it won't be always executed and assign resolver.ctx to a non-nil value.

File: central/graphql/resolvers/node_scan.go (hunk 15 above)
Comment (charmik-redhat) [reply central/graphql/resolvers/node_scan.go:44]: nodeScanResolver is created by `Scan()` sub-resolver is nodes graphQL and that sub-resolver adds the node's scope to nodeScanResolver's ctx before returning. So, nodeScanResolver should have a non-nil ctx at this point. But I think it is safe to add the check here in case the `Scan` sub-resolver changes.

File: central/graphql/resolvers/loaders/nodes.go [hunk 17]
```diff
@@ -58,16 +60,38 @@ type nodeLoaderImpl struct {
...
+	return nodes[0], nil
+}
+
+// FullNodeWithID loads full node from an ID.
+func (ndl *nodeLoaderImpl) FullNodeWithID(ctx context.Context, id string) (*storage.Node, error) {
+	node, err := ndl.FromID(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	// Load the full node if full scan is not available.
```
Comment (md2119) [inline central/graphql/resolvers/loaders/nodes.go:85]: The comment does not describe the if block below. A converse comment would work or move the comment after the if block.

File: central/graphql/resolvers/nodes.go [hunk 18]
```diff
@@ -519,7 +523,23 @@ func (resolver *nodeResolver) PlottedNodeVulnerabilities(ctx context.Context, ar
 }
 
 func (resolver *nodeResolver) Scan(ctx context.Context) (*nodeScanResolver, error) {
-	res, err := resolver.root.wrapNodeScan(resolver.data.GetScan(), true, nil)
+	// If Postgres is not enabled, node loader always pulls full node along with scan
+	scan := resolver.data.GetScan()
+	if env.PostgresDatastoreEnabled.BooleanSetting() {
+		// If scan is pulled, it is most likely to fetch all components and vulns contained in node.
```
Comment (md2119) [inline central/graphql/resolvers/nodes.go:529]: `...it is most likely for the user to fetch...`

**Review by md2119 — APPROVED** [review]
Nice job in getting this done!

**Review by keyallis — APPROVED** [review]
LGTM
